x/sponsorship/keeper/invariants.go

```diff
@@ -1,21 +1,135 @@
 package keeper
 
 import (
-	sdk "github.com/cosmos/cosmos-sdk/types"
+	"errors"
+	"fmt"
 
+	"cosmossdk.io/collections"
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension/v3/utils/uinv"
 	"github.com/dymensionxyz/dymension/v3/x/sponsorship/types"
 )
 
-func RegisterInvariants(ir sdk.InvariantRegistry, keeper Keeper) {
-	ir.RegisterRoute(types.ModuleName, "gauge-weights", GaugeWeightsInvariant(keeper))
+var invs = uinv.NamedFuncsList[Keeper]{
+	{Name: "delegator-validator-power", Func: InvariantDelegatorValidatorPower},
+	{Name: "distribution", Func: InvariantDistribution},
+	{Name: "votes", Func: InvariantVotes},
+	{Name: "general", Func: InvariantGeneral},
+}
+
+func RegisterInvariants(ir sdk.InvariantRegistry, k Keeper) {
+	invs.RegisterInvariants(types.ModuleName, ir, k)
+}
+
+// DO NOT DELETE
+func AllInvariants(k Keeper) sdk.Invariant {
+	return invs.All(types.ModuleName, k)
 }
 
-func GaugeWeightsInvariant(Keeper) sdk.Invariant {
-	// TODO
-	return func(ctx sdk.Context) (string, bool) {
-		var broken bool
-		var msg string
+func InvariantDelegatorValidatorPower(k Keeper) uinv.Func {
+	return uinv.AnyErrorIsBreaking(func(ctx sdk.Context) error {
+		return k.delegatorValidatorPower.Walk(ctx, nil,
+			func(key collections.Pair[sdk.AccAddress, sdk.ValAddress], value math.Int) (stop bool, err error) {
+				if value.IsNegative() {
+					return false, fmt.Errorf("negative power: %s", value)
+				}
+				return false, nil
+			})
+	})
+}
+
+func InvariantDistribution(k Keeper) uinv.Func {
+	return uinv.AnyErrorIsBreaking(func(ctx sdk.Context) error {
+		d, err := k.GetDistribution(ctx)
+		if err != nil {
+			return fmt.Errorf("get distribution: %w", err)
+		}
+
+		t := math.ZeroInt()
+		for _, g := range d.Gauges {
+			t = t.Add(g.Power)
+		}
+		if t.GT(d.VotingPower) {
+			return fmt.Errorf("sum of gauge power exceeds total power sum: %s: total: %s", t, d.VotingPower)
+		}
+
+		var errs []error
+		for _, g := range d.Gauges {
+			if g.Power.IsNegative() {
+				errs = append(errs, fmt.Errorf("negative g power: %d: %s", g.GaugeId, g.Power))
+			}
+		}
+		if len(errs) > 0 {
+			return errors.Join(errs...)
+		}
+
+		if d.VotingPower.IsNegative() {
+			return fmt.Errorf("negative total voting power: %s", d.VotingPower)
+		}
+
+		return nil
+	})
+}
+
+func InvariantVotes(k Keeper) uinv.Func {
+	return uinv.AnyErrorIsBreaking(func(ctx sdk.Context) error {
+		err := k.IterateVotes(ctx, func(voter sdk.AccAddress, vote types.Vote) (stop bool, err error) {
+			if vote.VotingPower.IsNegative() {
+				return false, fmt.Errorf("negative voting power: %s", vote.VotingPower)
+			}
+			t := sdk.ZeroInt()
+			for _, weight := range vote.GetWeights() {
+				w := weight.Weight
+				if w.LT(sdk.OneInt()) || w.GT(sdk.NewInt(100)) {
```

Comment:
in sponsorship we use different units for %. plus lower bound may be <1%, so i'd the following
```suggestion
				if !w.IsPositive() || w.GT(types.MaxAllocationWeight) {
```
this is how gauge weight is validated rn https://github.com/dymensionxyz/dymension/blob/a7a03787193ecdfc8d3d1add134e31f0512c74b2/x/sponsorship/types/types.go#L54-L62

Reply:
btw have you tried to run these invariants somehow?

Reply:
anyways fixed the weight